tests/unit/test_unit_smhi.py

```diff
@@ -0,0 +1,249 @@
+    @patch("smhi.metobs.Stations.__new__")
+    def test_unit_smhi_get_stations(self, mock_station_data, parameter):
+        """Unit test for SMHI get_stations method.
+
+        Args:
+            parameter: parameter (int)
+        """
+        client = SMHI()
+        assert client.get_stations(parameter)
+
+    @pytest.mark.parametrize("parameter_title", [(None), ("Snöfall")])
+    @patch("smhi.metobs.Stations.__new__")
+    def test_unit_smhi_get_stations_from_title(
+        self, mock_station_data, parameter_title
+    ):
+        """Unit test for SMHI get_stations_from_title method.
+
+        Args:
+            title: title of station
+        """
+        client = SMHI()
+        assert client.get_stations_from_title(parameter_title)
+
+    @pytest.mark.parametrize(
+        "parameter, station, distance", [(8, 180960, None), (8, 180960, 50)]
+    )
+    @patch("smhi.metobs.Stations.__new__")
+    @patch("smhi.metobs.Periods.__new__")
+    @patch("smhi.metobs.Data.__new__")
+    @patch("smhi.smhi.SMHI._interpolate")
+    def test_unit_get_data(
+        self,
+        mock_interpolate,
+        mock_data_data,
+        mock_period_data,
+        mock_station_data,
+        parameter,
+        station,
+        distance,
+    ):
+        mock_interpolate.return_value = "test"
+        client = SMHI()
+        data = client.get_data(parameter, station, distance)
+        assert data == "test"
+
+    @pytest.mark.parametrize(
+        "parameter, city, distance", [(8, "Bengtsfors", None), (8, "Bengtsfors", 50)]
+    )
+    @patch("smhi.metobs.Stations.__new__")
+    @patch("smhi.metobs.Periods.__new__")
+    @patch("smhi.metobs.Data.__new__")
+    @patch("smhi.smhi.SMHI._find_stations_by_city")
+    @patch("smhi.smhi.SMHI._interpolate")
+    def test_unit_get_data_by_city(
+        self,
+        mock_interpolate,
+        mock_find_stations_by_city,
+        mock_data_data,
+        mock_period_data,
+        mock_station_data,
+        parameter,
+        city,
+        distance,
+    ):
+        mock_interpolate.return_value = "test"
+        client = SMHI()
+        data = client.get_data_by_city(parameter, city, distance)
+        assert data == "test"
+
+    distanceresponse = MagicMock()
+    distanceresponse.km = 0
+
+    @pytest.mark.parametrize(
+        "latitude, longitude,dist",
+        [
+            (
+                59,
+                17,
+                0,
+            ),
+            (
+                59.4,
+                17,
+                30,
+            ),
+        ],
+    )
+    @patch("geopy.distance.distance", return_value=distanceresponse)
+    def test_find_stations_from_gps(self, mock_distance, latitude, longitude, dist):
+        """Unit test for SMHI find_stations_from_gps method.
+
+        Args:
+        """
+        station1 = MagicMock()
+        station1.id = 1
+        station1.name = "Akalla"
+        station1.latitude = 59.5
+        station1.longitude = 17.8
```

Comment:
Are these provided as parameters?

Reply:
I was not able to set the system up where the mocked "return_value" stuff worked by actually returning what I wanted, so I ended up with this solution. If you want me to do it with an input, show me how it is done. I was unable to get it from stack overflow visits over two days.

Reply:
I guess one could define two classes at the top of the file before the test class, like this

```
class MockMetobsStationLink:
    def __init__(self, id, name, lat, lon):
        self.id = id
        self.name = name
        self.latitude = lat
        self.longitude = lon


class MockParameterModel:
    def __init__(self, station):
        self.station = station
```

to easier add more stations. Nothing wrong with the `MagicMock` solution you have, just a bit harder to extend. Something like this maybe (removing `@patch("geopy.distance.distance", return_value=distanceresponse)` so we just use that function in the test)

```python
@pytest.mark.parametrize(
    "latitude, longitude, dist, expected_result", [(59, 17, 0, RES1), (59.4, 17, 30, RES2)]
)
def test_find_stations_from_gps(self, mock_distance, latitude, longitude, dist, expected_result):
    station1 = MockMetobsStationLink(1, "name", 50, 16)
    station2 = MockMetobsStationLink(1, "name", 50, 16)
    station3 = MockMetobsStationLink(1, "name", 50, 16)
    stations = MockParameterModel([station1, station2, station3])

    ....
```

with `RES1` and ´RES2´ whatever result you expect.

Note that, this whole block would be perfect for a `pytest.fixture`

```python
    station1 = MockMetobsStationLink(1, "name", 50, 16)
    station2 = MockMetobsStationLink(1, "name", 50, 16)
    station3 = MockMetobsStationLink(1, "name", 50, 16)
    stations = MockParameterModel([station1, station2, station3])
```

Reply:
I will give this a shot. Thanks for a good description!